pkg/restore/merge_service_account.go

```diff
@@ -0,0 +1,135 @@
+// Labels and Annotations on the back up but not on the in-cluster copy will be merged. If a key is specified in both, the in-cluster version is retained.
+func mergeServiceAccounts(fromCluster, fromBackup *unstructured.Unstructured) (*unstructured.Unstructured, error) {
+	// merge any labels, all secrets except the default token, and image pull secrets from the backup
+	desired := new(corev1api.ServiceAccount)
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(fromCluster.UnstructuredContent(), desired); err != nil {
+		return nil, errors.Wrap(err, "unable to convert from-cluster service account from unstructured to serviceaccount")
+
+	}
+
+	backupSA := new(corev1api.ServiceAccount)
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(fromBackup.UnstructuredContent(), backupSA); err != nil {
+		return nil, errors.Wrap(err, "unable to convert from backed up service account unstructured to serviceaccount")
+
+	}
+
+	for i := len(backupSA.Secrets) - 1; i >= 0; i-- {
+		secret := &backupSA.Secrets[i]
+		if strings.HasPrefix(secret.Name, "default-token-") {
+			// Copy all secrets *except* default-token
+			backupSA.Secrets = append(backupSA.Secrets[:i], backupSA.Secrets[i+1:]...)
+			break
+		}
+	}
+
+	desired.Secrets = mergeObjectReferenceSlices(desired.Secrets, backupSA.Secrets)
+
+	desired.ImagePullSecrets = mergeLocalObjectReferenceSlices(desired.ImagePullSecrets, backupSA.ImagePullSecrets)
+
+	collections.MergeMaps(desired.Labels, backupSA.Labels)
+
+	collections.MergeMaps(desired.Annotations, backupSA.Annotations)
+
+	desiredUnstructured, err := runtime.DefaultUnstructuredConverter.ToUnstructured(desired)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to convert desired service account to unstructured")
+	}
+
+	return &unstructured.Unstructured{Object: desiredUnstructured}, nil
+}
+
+func mergeObjectReferenceSlices(first, second []corev1api.ObjectReference) []corev1api.ObjectReference {
+
+	for _, s := range second {
+		var exists bool
+		for _, f := range first {
+			// if e, err := equal; if e && err == nil
+			if s.Name == f.Name {
+				exists = true
+				break
+			}
+		}
+		if !exists {
+			first = append(first, s)
+		}
+	}
+
+	return first
+}
+
+func mergeLocalObjectReferenceSlices(first, second []corev1api.LocalObjectReference) []corev1api.LocalObjectReference {
+
+	for _, s := range second {
+		var exists bool
+		for _, f := range first {
+			// if e, err := equal; if e && err == nil
+			if s.Name == f.Name {
+				exists = true
+				break
+			}
+		}
+		if !exists {
+			first = append(first, s)
+		}
+	}
+
+	return first
+}
+
+// generatePatch will calculate a JSON merge patch for an object's desired state.
+// If the passed in objects are already equal, nil is returned.
+func generatePatch(fromCluster, desired *unstructured.Unstructured) ([]byte, error) {
+	equal, err := objectsAreEqual(fromCluster, desired)
```

Comment:
`objectsAreEqual` mutates `fromCluster`. This probably isn't a good thing here. It might be better to create a custom `conversion.Equalities` instance that has equality functions specific for Ark's needs that don't require mutations.

Reply:
Thanks - I wasn't super happy with it mutating, either. I'll look at writing a custom function